Updated the building connectors documentation (#80)
ConduitIO/conduit-site: docs/connectors/building-connectors/developing-destination-connectors.mdx

---
title: "Developing a Destination Connector"
sidebar_position: 4
---

A Destination is responsible for writing [Record](https://pkg.go.dev/github.com/conduitio/conduit-connector-sdk#Record) to third party systems.

You need to implement the functions required by Destination and provide your own implementations. Information about individual functions are listed below.

## `destination.go`

This file provides the main functionality of your Destination Connector.

> - **`Destination`** Struct: Every Destination implementation needs to include an [UnimplementedDestination](https://pkg.go.dev/github.com/conduitio/conduit-connector-sdk#UnimplementedDestination) to satisfy the interface. This allows us to potentially change the interface in the future while remaining backwards compatible with existing Destination implementations.  This struct can be modified to add additional fields that can be accessed throughout the lifecycle of the Connector

Reviewer: ```suggestion
- **`Destination`** Struct: Every Destination implementation needs to include an [UnimplementedDestination](https://pkg.go.dev/github.com/conduitio/conduit-connector-sdk#UnimplementedDestination) to satisfy the interface. This allows us to potentially change the interface in the future while remaining backward compatible with existing Destination implementations.  This struct can be modified to add additional fields that can be accessed throughout the lifecycle of the Connector.
```

    ```go
    type Destination struct {
        sdk.UnimplementedDestination

        config DestinationConfig
    }
    ```

> - **`NewDestination()`**: A constructor function for your Destination struct. Note that this is the same function that should be set as the value of Connector.NewDestination. The constructor should be used to wrap your Destination in the default middleware. You can add additional middleware, but unless you have a very good reason, you should always include the default middleware.

Reviewer: ```suggestion
- **`NewDestination()`**: A constructor function for your Destination struct. Note that this is the same function that should be set as the value of `Connector.NewDestination`. The constructor should be used to wrap your Destination in the default middleware. You can add additional middleware, but unless you have a very good reason, you should always include the default middleware.
```

    ```go
    func NewDestination() sdk.Destination {
        // Create Destination and wrap it in the default middleware.
        return sdk.DestinationWithMiddleware(&Destination{}, sdk.DefaultDestinationMiddleware()...)
    }
    ```

> - **`Parameters()`**: A map of named Parameters that describe how to configure the connector. This map is typically generate using `paramgen`

Reviewer: ```suggestion
- **`Parameters()`**: A map of named Parameters that describe how to configure the connector. This map is typically generated using [`paramgen`](https://github.com/ConduitIO/conduit-commons/tree/main/paramgen).
```

    ```go
    func (d *Destination) Parameters() map[string]sdk.Parameter {
        return d.config.Parameters()
    }
    ```

- **`Configure()`**: Validates and stores configuration data for the connector. Any complex validation logic should be implemented here.

    ```go
    func (d *Destination) Configure(ctx context.Context, cfg map[string]string) error {
        err := sdk.Util.ParseConfig(cfg, &d.config)
        if err != nil {
            return fmt.Errorf("invalid config: %w", err)

>         }

Reviewer: ```suggestion
        }
        // custom validations here
```

        return nil
    }
    ```

- **`Open()`**: Prepares the connector to start producing records based on the last known successful position. If needed, the connector should open connections in this function.

    ```go
    func (d *Destination) Open(ctx context.Context) error {
        // Retrieve the directory path from the config
        directoryPath := d.config.Directory

        // Check if the directory exists
        if _, err := os.Stat(directoryPath); os.IsNotExist(err) {
            // Create the directory if it doesn't exist
            err := os.MkdirAll(directoryPath, 0755)
            if err != nil {
                return fmt.Errorf("failed to create directory '%s': %w", directoryPath, err)
            }
        } else if err != nil {
            // Return any error other than the directory not existing
            return fmt.Errorf("error checking directory '%s': %w", directoryPath, err)
        }

        // The directory exists (or was just created), so we can proceed
        return nil
    }
    ```

> - **`Write()`**: Writes len(records) from a slice of sdk.Record objects received from the Conduit pipeline to the destination right away without caching.  It should return the number of records written from the slice and any error encountered that caused the write to stop early.

Reviewer: ```suggestion
- **`Write()`**: Writes len(records) from a slice of `sdk.Record` objects received from the Conduit pipeline to the destination right away without caching.  It should return the number of records written from the slice and any error encountered that caused the write to stop early.
```
